### ISSUE TITLE
Stackdriver queries with custom Group By not supported by Alias By

### ISSUE THREAD

**Issue by jcxplorer** [grafana/grafana#15552]
### What Grafana version are you using?

v6.0.0-beta3

### What datasource are you using?

Stackdriver

### What OS are you running grafana on?

Linux (official docker image)

### What did you do?

Added the following query, with a custom _Group By_ value. This value wasn't available from the dropdown options, but is valid and works in Stackdriver's metric explorer.

<img width="621" alt="screenshot 2019-02-20 at 17 36 13" src="https://user-images.githubusercontent.com/42855/53104681-7f32ff00-3538-11e9-945f-3ffbfdf649d7.png">

### What was the expected result?

The _Alias By_ value works and shows the service names in the graph.

### What happened instead?

No service names are shown, shows `{{metadata.system_labels.service_name}}` instead.

<img width="959" alt="screenshot 2019-02-20 at 17 45 37" src="https://user-images.githubusercontent.com/42855/53104051-5a8a5780-3537-11e9-903a-ab3fa344bd95.png">

### Raw network request & response (from Query Inspector)

```js
{
  "xhrStatus": "complete",
  "request": {
    "method": "POST",
    "url": "api/tsdb/query",
    "data": {
      "from": "1550666795187",
      "to": "1550677595187",
      "queries": [
        {
          "refId": "A",
          "intervalMs": 15000,
          "datasourceId": 4,
          "metricType": "kubernetes.io/container/cpu/limit_utilization",
          "crossSeriesReducer": "REDUCE_SUM",
          "perSeriesAligner": "ALIGN_MEAN",
          "alignmentPeriod": "stackdriver-auto",
          "groupBys": [
            "metadata.system_labels.service_name"
          ],
          "view": "FULL",
          "filters": [
            "resource.label.namespace_name",
            "=",
            "ingress"
          ],
          "aliasBy": "{{metadata.system_labels.service_name}}",
          "type": "timeSeriesQuery"
        }
      ]
    }
  },
  "response": {
    "results": {
      "A": {
        "refId": "A",
        "meta": {
          "alignmentPeriod": 60,
          "groupBys": [
            "metadata.system_labels.service_name"
          ],
          "metricLabels": {},
          "rawQuery": "aggregation.alignmentPeriod=%2B60s&aggregation.crossSeriesReducer=REDUCE_SUM&aggregation.groupByFields=metadata.system_labels.service_name&aggregation.perSeriesAligner=ALIGN_MEAN&filter=metric.type%3D%22kubernetes.io%2Fcontainer%2Fcpu%2Flimit_utilization%22+resource.label.namespace_name%3D%22ingress%22&interval.endTime=2019-02-20T15%3A46%3A35Z&interval.startTime=2019-02-20T12%3A46%3A35Z&view=FULL",
          "resourceLabels": {
            "project_id": [
              "[REDACTED]"
            ]
          },
          "resourceTypes": [
            "k8s_container"
          ]
        },
        "series": [
          {
            "name": "{{metadata.system_labels.service_name}}",
            "points": [
              [
                0.005439716072260126,
                1550666855000
              ],
              [
                0.005644686000020253,
                1550666915000
              ],
              // more...
           }
        ],
        "tables": null
      }
    }
  }
}
```

**Comment by daniellee** [comment]
Thank you for the well-written issue. There are two parts to this:

- can we get a list of custom groupbys from the Stackdriver API to show in the groupby dropdown
- fix the alias by function to look at custom metadata fields

**Comment by jcxplorer** [comment]
Looking at the network tab while using Stackdriver, I was able to see it gets the data with:

    GET https://content-stackdriver.googleapis.com/v1beta2/projects/[REDACTED]/resourceMetadataHistogram?filter=resource.type%20%3D%20%22k8s_container%22&%24unique=gc764

Response looks like:

```json
{
  "histogram": [
    {
      "facetId": "resource.labels",
      "values": [
        {
          "facetValueId": "cluster_name"
        },
        {
          "facetValueId": "container_name"
        },
        {
          "facetValueId": "location"
        },
        {
          "facetValueId": "namespace_name"
        },
        {
          "facetValueId": "pod_name"
        }
      ]
    },
    {
      "facetId": "metadata.system_labels",
      "values": [
        {
          "facetValueId": "state"
        },
        {
          "facetValueId": "container_image"
        },
        {
          "facetValueId": "container_image_id"
        },
        {
          "facetValueId": "name"
        },
        {
          "facetValueId": "node_name"
        },
        {
          "facetValueId": "provider_instance_id"
        },
        {
          "facetValueId": "provider_resource_type"
        },
        {
          "facetValueId": "provider_zone"
        },
        {
          "facetValueId": "service_name"
        },
        {
          "facetValueId": "top_level_controller_name"
        },
        {
          "facetValueId": "top_level_controller_type"
        }
      ]
    },
    {
      "facetId": "metadata.user_labels",
      "values": [
        {
          "facetValueId": "alertmanager"
        },
        {
          "facetValueId": "app"
        },
        {
          "facetValueId": "chart"
        },
        {
          "facetValueId": "component"
        },
        {
          "facetValueId": "controller-revision-hash"
        },
        {
          "facetValueId": "controller-uid"
        },
        {
          "facetValueId": "heritage"
        },
        {
          "facetValueId": "job-name"
        },
        {
          "facetValueId": "k8s-app"
        },
        {
          "facetValueId": "kubernetes.io/cluster-service"
        },
        {
          "facetValueId": "name"
        },
        {
          "facetValueId": "operator"
        },
        {
          "facetValueId": "pod-template-generation"
        },
        {
          "facetValueId": "pod-template-hash"
        },
        {
          "facetValueId": "prometheus"
        },
        {
          "facetValueId": "release"
        },
        {
          "facetValueId": "role"
        },
        {
          "facetValueId": "statefulset.kubernetes.io/pod-name"
        },
        {
          "facetValueId": "tier"
        },
        {
          "facetValueId": "version"
        }
      ]
    }
  ],
  "totalSize": 36
}
```

**Comment by daniellee** [comment]
Will see if I can squeeze this into 6.0 at the last minute.

**Comment by daniellee** [comment]
Will investigate some more but it looks like resourceMetadataHistogram isn't part of the public API. Stackdriver has some internal api endpoints which we can't use. Will see if it possible to get this metadata another way.

**Comment by daniellee** [comment]
So it does work if you enter the filter manually:

![image](https://user-images.githubusercontent.com/434655/53286485-a9bacd00-376f-11e9-99f3-b0ca44c82e59.png)

**Comment by jcxplorer** [comment]
Yes, entering it manually for either filtering or grouping does work. I think that’s good enough even if autocomplete doesn’t work.

The dealbreaker I see is the lack of support in the alias by function. Fixing that alone would be amazing 👍

**Comment by daniellee** [comment]
Tested some more and in the Stackdriver UI, the response contains a meta.systemLabels object but this is not part of the response when using the public API. So I'm stuck on this until I get help from the Stackdriver team.

**Comment by cbandy** [comment]
I, too, would be happy without dropdowns if I could at least alias.

@daniellee which public API? When I use [this API explorer](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.timeSeries/list), it makes calls to `https://content-monitoring.googleapis.com/v3/projects/[REDACTED]/timeSeries` and I can see my container labels in `metadata.userLabels`.

For example, with
`aggregation.groupByFields=metadata.user_labels.dag_id&aggregation.groupByFields=metadata.user_labels.task_id`

I receive the following payload

```json
{
  "timeSeries": [
    {
      "metric": {
        "type": "container.googleapis.com/container/cpu/usage_time"
      },
      "resource": {
        "type": "gke_container",
        "labels": {
          "namespace_id": "...",
          "instance_id": "...",
          "zone": "...",
          "pod_id": "...",
          "project_id": "...",
          "cluster_name": "...",
          "container_name": "..."
        }
      },
      "metricKind": "GAUGE",
      "valueType": "DOUBLE",
      "points": [
        {
          "interval": {
            "startTime": "2019-03-19T16:06:01Z",
            "endTime": "2019-03-19T16:06:01Z"
          },
          "value": {
            "doubleValue": 0.011390183361388927
          }
        }
      ],
      "metadata": {
        "userLabels": {
          "task_id": "xyz",
          "dag_id": "abc"
        }
      }
    },
    ...
  ]
}
```

**Comment by cbandy** [comment]
I know this isn't _directly_ related, but some news has come about GKE ***beta*** _log_ monitoring:

https://cloud.google.com/kubernetes-engine/docs/release-notes#april_25_2019
https://cloud.google.com/monitoring/kubernetes-engine/release-guide#ki1_12_6

> Stackdriver Kubernetes Monitoring logging label fields change when you upgrade your GKE clusters to GKE v1.12.6 or higher.
> * Kubernetes Pod labels, currently located in the `metadata.userLabels` field, are moved to the `labels` field in the [LogEntry][] and the label keys have a prefix prefix of `k8s-pod/`.
> * Stackdriver system labels that are in the `metadata.systemLabels` field are no longer available.

[LogEntry]: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry

**Comment by ScubaDrew** [comment]
@cbandy indicated the public apis return the metadata. Being able to use the metadata in the alias would be amazing.

**Comment by auvideus** [comment]
@danielle metadata labels (both system and user) are returned by the public API, but only the ones used in the `aggregation.groupByFields` field, as can be seen in cbandy's post above.

https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries

![Screen Shot 2019-07-10 at 2 45 16 PM](https://user-images.githubusercontent.com/1489873/60999699-65e0f280-a321-11e9-9c49-699f710ef66e.png)

It would be great if that data was used to make Alias By function correctly for metadata labels used in Group By.  A lot of Stackdriver metrics, particularly the [agent metrics](https://cloud.google.com/monitoring/api/metrics_agent), don't include instance_name as a metric label, and therefore you have to rely on getting it via the built-in metadata label.  Without Alias By, the best I can do is display the instance_id (a resource label) which is a lot less useful.

**Comment by daniellee** [comment]
@auvideus you mentioned the wrong person someone called Danielle got pinged. (My name is Daniel Lee or @daniellee)

I'll have another a look when I looked last time at how the Stackdriver UI got that metadata it was through an internal API and I did not see the custom fields in the public API when I tested. But maybe I missed something.

**Comment by daniellee** [comment]
Closed in favour of #18887 which describes the issue better.
